extensions/oidc/runtime/src/main/java/io/quarkus/oidc/runtime/OidcAuthenticationMechanism.java

```diff
@@ -0,0 +1,53 @@
+package io.quarkus.oidc.runtime;
+
+import java.util.concurrent.CompletionStage;
+
+import javax.enterprise.context.ApplicationScoped;
+import javax.inject.Inject;
+
+import io.quarkus.security.identity.IdentityProviderManager;
+import io.quarkus.security.identity.SecurityIdentity;
+import io.quarkus.vertx.http.runtime.security.ChallengeData;
+import io.quarkus.vertx.http.runtime.security.HttpAuthenticationMechanism;
+import io.vertx.ext.web.RoutingContext;
+
+@ApplicationScoped
+public class OidcAuthenticationMechanism implements HttpAuthenticationMechanism {
+
+    @Inject
+    DefaultTenantConfigResolver resolver;
+    private volatile BearerAuthenticationMechanism bearerAuth;
+    private volatile CodeAuthenticationMechanism codeAuth;
+
+    @Override
+    public CompletionStage<SecurityIdentity> authenticate(RoutingContext context,
+            IdentityProviderManager identityProviderManager) {
+        return isWebApp(context) ? getCodeAuth().authenticate(context, identityProviderManager, resolver)
+                : getBearerAuth().authenticate(context, identityProviderManager, resolver);
+    }
+
+    @Override
+    public CompletionStage<ChallengeData> getChallenge(RoutingContext context) {
+        return isWebApp(context) ? getCodeAuth().getChallenge(context, resolver)
+                : getBearerAuth().getChallenge(context, resolver);
+    }
+
+    private CodeAuthenticationMechanism getCodeAuth() {
+        if (codeAuth == null) {
```

Comment:
Not that it matters, but this can actually create multiple instances here if multiple threads hit it at the same time.

Reply:
Yes, this is what I was typing about [above](https://github.com/quarkusio/quarkus/pull/7086#issuecomment-584300029) but I decided to give it a try to make sure for a single `application-type` no code (or bearer) flow is created given Pedro's concern.
I'm going to return to the original code where I was just creating these 2 instances.
With the dynamic `TenantConfigResolver` one can never be 100% sure there will only be one application-type